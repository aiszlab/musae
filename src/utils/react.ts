import { Partialable } from "@aiszlab/relax/types";
import { Children, isValidElement, type ReactNode } from "react";

interface ElementProps {
  children?: ReactNode;
}

/**
 * @description
 * get the string of component children
 */
export const toReactNodeText = (children: ReactNode): Partialable<string> => {
  const _children = Children.toArray(children);

  for (const child of _children) {
    if (typeof child === "string" || typeof child === "number") {
      return child.toString();
    }

    if (isValidElement<ElementProps>(child)) {
      const textInChild = toReactNodeText(child.props.children);

      if (textInChild) {
        return textInChild;
      }
    }
  }
};
