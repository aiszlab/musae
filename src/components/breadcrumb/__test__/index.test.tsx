import { render } from "@testing-library/react";
import Breadcrumb from "../breadcrumb";
import React from "react";
import "@testing-library/jest-dom";

describe("`Breadcrumb` Component", () => {
  test("empty items", () => {
    const { container } = render(<Breadcrumb items={[]} />);
    expect(container.hasChildNodes()).toBeFalsy();
  });

  test("display items", () => {
    const { container } = render(
      <Breadcrumb items={[{ label: "label1" }, { label: "link2", href: "https://aiszlab.dev" }, { label: "label3" }]} />
    );

    expect(container.hasChildNodes()).toBeTruthy();
    expect(container.querySelectorAll(".musae-breadcrumb__item").length).toBe(3);
    expect(container.querySelectorAll(".musae-breadcrumb__separator").length).toBe(2);
  });

  test("customize separator", () => {
    const { getAllByRole } = render(
      <Breadcrumb
        items={[{ label: "label1" }, { label: "link2", href: "https://aiszlab.dev" }, { label: "label3" }]}
        separator="-"
      />
    );

    expect(getAllByRole("separator").filter((item) => item.innerHTML === "-").length).toBe(2);
  });
});
