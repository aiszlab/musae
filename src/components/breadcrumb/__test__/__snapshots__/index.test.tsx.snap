// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Breadcrumb\` Component customize separator 1`] = `
<DocumentFragment>
  <nav
    class="musae-breadcrumb theme__label.large breadcrumb__styles.breadcrumb"
    style="--color-on-surface-variant: #49454F;"
  >
    <ol
      class="breadcrumb__styles.navigations"
    >
      <li
        class="musae-breadcrumb__item item__default"
        style="--color-surface: #FFFBFE; --color-on-surface: #1D1B20;"
      >
        label1
      </li>
      <li
        class="musae-breadcrumb__separator item__default"
        role="separator"
      >
        -
      </li>
      <li
        class="musae-breadcrumb__item item__default item__link"
        style="--color-surface: #FFFBFE; --color-on-surface: #1D1B20;"
      >
        <a
          href="https://aiszlab.dev"
        >
          link2
        </a>
      </li>
      <li
        class="musae-breadcrumb__separator item__default"
        role="separator"
      >
        -
      </li>
      <li
        class="musae-breadcrumb__item item__default"
        style="--color-surface: #FFFBFE; --color-on-surface: #1D1B20;"
      >
        label3
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;

exports[`\`Breadcrumb\` Component display items 1`] = `
<DocumentFragment>
  <nav
    class="musae-breadcrumb theme__label.large breadcrumb__styles.breadcrumb"
    style="--color-on-surface-variant: #49454F;"
  >
    <ol
      class="breadcrumb__styles.navigations"
    >
      <li
        class="musae-breadcrumb__item item__default"
        style="--color-surface: #FFFBFE; --color-on-surface: #1D1B20;"
      >
        label1
      </li>
      <li
        class="musae-breadcrumb__separator item__default"
        role="separator"
      >
        /
      </li>
      <li
        class="musae-breadcrumb__item item__default item__link"
        style="--color-surface: #FFFBFE; --color-on-surface: #1D1B20;"
      >
        <a
          href="https://aiszlab.dev"
        >
          link2
        </a>
      </li>
      <li
        class="musae-breadcrumb__separator item__default"
        role="separator"
      >
        /
      </li>
      <li
        class="musae-breadcrumb__item item__default"
        style="--color-surface: #FFFBFE; --color-on-surface: #1D1B20;"
      >
        label3
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;

exports[`\`Breadcrumb\` Component empty items 1`] = `<DocumentFragment />`;
